src/web/app/user-join-page.component.ts

```diff
@@ -43,37 +47,70 @@ export class UserJoinPageComponent implements OnInit {
       this.key = queryParams.key;
       this.institute = queryParams.instructorinstitution;
       this.mac = queryParams.mac;
+      this.isCreatingAccount = queryParams.iscreatingaccount === 'true';
 
-      if (this.institute != null && this.mac == null) {
+      if (this.key == null || (this.institute != null && this.mac == null)) {
         this.validUrl = false;
         return;
       }
 
-      this.courseService.getJoinCourseStatus(this.key, this.entityType).subscribe((resp: JoinStatus) => {
-        this.hasJoined = resp.hasJoined;
-        this.userId = resp.userId || '';
-        if (this.hasJoined && this.userId) {
-          // The regkey has been used and there is a logged in user.
-          // Simply redirect the user to their home page, regardless of whether the regkey matches or not.
-          this.navigationService.navigateByURL(this.router, `/web/${this.entityType}/home`);
-        } else {
-          this.isLoading = false;
-        }
-      }, (resp: ErrorMessageOutput) => {
-        if (resp.status === 403) {
-          this.isLoading = false;
-          const nextUrl: string = `${window.location.pathname}${window.location.search.replace(/&/g, '%26')}`;
-          this.authService.getAuthUser(undefined, nextUrl).subscribe((auth: AuthInfo) => {
-            if (!auth.user) {
-              window.location.href = `${this.backendUrl}${auth.studentLoginUrl}`;
-            }
-          });
-        } else {
-          const modalRef: any = this.ngbModal.open(ErrorReportComponent);
-          modalRef.componentInstance.requestId = resp.error.requestId;
-          modalRef.componentInstance.errorMessage = resp.error.message;
-        }
-      });
+      if (this.isCreatingAccount) {
+        this.setupForCreateAccount();
+      } else {
+        this.setupForJoinCourse();
+      }
+    });
+  }
+
+  /**
+   * Setup page for instructor to join course.
+   */
+  setupForJoinCourse(): void {
+    this.courseService.getJoinCourseStatus(this.key, this.entityType).subscribe((resp: JoinStatus) => {
```

Comment:
This part has been refactored in #11421. You probably would find even more common operations/steps with the create account counterpart.